File: hw02_happy_tickets/hw02_mid.py
# Нахождение кол-ва счастливых билетов 2N с помощью рекурсии
# Для уровня Middle

def count_happy_tickets(n):
    sum1 = 0
    sum2 = 0
    count = 0
    
    def count_tickets(n, sum1, sum2):
        nonlocal count
        if n == 0:
            if sum1 == sum2:
               count += 1
            return
        for n1 in range(0, 10):
            for n2 in range(0, 10):
                count_tickets(n-1, sum1+n1, sum2+n2)
            
    count_tickets(n, 0, 0)
    return count
    
def main():
    n = int(input("Введите кол-во цифр: "))
    count = count_happy_tickets(n)
    print(count)

if __name__ == "__main__":
    main()

File: hw02_happy_tickets/hw02_mid.cpp
#include <iostream>

unsigned long long int count = 0;

int happy_count(int digits, int sum1, int sum2)
{
    if (digits == 0)
    {
        if (sum1 == sum2)
        {
            count++;
        }
        return 0;
    }
    for (int d1=0; d1<10; d1++)
        for (int d2=0; d2<10; d2++)
            happy_count(digits-1, sum1+d1, sum2+d2);
    return 0;
}

int main()
{
    int digits;
    std::cout << "Введите количество цифр = ";
    std::cin >> digits;
    int sum1 = 0;
    int sum2 = 0;
    happy_count(digits, sum1, sum2);
    std::cout << count << std::endl;
    return 0;
}


